plugins/ccgx/fu-ccgx-hpi-device.c

```diff
@@ -1454,16 +1495,6 @@ fu_ccgx_hpi_device_init (FuCcgxHpiDevice *self)
 	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
 	fu_device_add_flag (FU_DEVICE (self), FWUPD_DEVICE_FLAG_SELF_RECOVERY);
 
-	/* we can recover the I²C link using reset */
-	fu_device_retry_add_recovery (FU_DEVICE (self),
-				      G_USB_DEVICE_ERROR,
-				      G_USB_DEVICE_ERROR_IO,
-				      fu_ccgx_hpi_device_reset_cb);
-	fu_device_retry_add_recovery (FU_DEVICE (self),
-				      G_USB_DEVICE_ERROR,
-				      G_USB_DEVICE_ERROR_TIMED_OUT,
-				      fu_ccgx_hpi_device_reset_cb);
-
```

Comment:
Why did you drop these?  The whole point of them was so taht for failures and resets that it could be fixed from callback functions instead.

Reply:
fu_ccgx_hpi_device_reset_i2c() can not be used to clear USB error but it is 
used to clear i2c error between USB serial device and PD IC.
i2c error can be found by calling fu_ccgx_hpi_device_get_i2c_status().

In fu_ccgx_hpi_device_i2c_write() and fu_ccgx_hpi_device_i2c_read(), we check i2c status and check i2c status to clear i2c error and retry i2c communication.

Reply:
Okay, go back a bit. To do, for instance, an i2c read we do:

 * A control transfer with CY_I2C_READ_CMD
 * A bulk transfer with the buffer
 * An interrupt transfer to get the notification buffer

If `CY_I2C_ERROR_BIT` is set in the notification buffer, we should according to you do an i2c reset. And then what do we retry? All three steps? Just the bulk transfer?

My main point is - if we want to do the i2c reset (and it seems like we should) we can't just fail, reset, then return failure. We need to fail, reset, then try again -- using fu_device_retry() rather than a hand-rolled loop.

Reply:
Raw i2c read and write is consist of control xfer, bulk xfer, intr xfer.

In fu_ccgx_hpi_device_reg_read(), I2C read operation need two steps : 
  - fu_ccgx_hpi_device_reg_read() {
         fu_ccgx_hpi_device_i2c_write()
         fu_ccgx_hpi_device_i2c_read()
   }

For i2c read operation retry, we need to restart two steps after sending fu_ccgx_hpi_device_reset_i2c() when CY_I2C_ERROR_BIT is set in fu_ccgx_hpi_device_i2c_write() or fu_ccgx_hpi_device_i2c_read().

In fu_ccgx_hpi_device_reg_write(), I2C Write operation need one steps: 
  - fu_ccgx_hpi_device_reg_write() {
       fu_ccgx_hpi_device_i2c_write()
    }

For i2c write operation retry, we need to restart one step after sending fu_ccgx_hpi_device_reset_i2c() when CY_I2C_ERROR_BIT is set in fu_ccgx_hpi_device_i2c_write() .

we can check CY_I2C_ERROR_BIT in both notification buffer (intr xfer) and fu_ccgx_hpi_device_get_i2c_status() result.

Could you revise my manual i2c operation retry using fu_device_retry()?
I don't know how to use it.

Reply:
> Could you revise my manual i2c operation retry using fu_device_retry

Give me a few mins.

Reply:
@chlee75 can you have a look at the https://github.com/fwupd/fwupd/pull/1959 PR please.

Reply:
I added my comment.
you need to retry fu_ccgx_hpi_device_reg_read() instead of fu_ccgx_hpi_device_i2c_read()
you need to retry fu_ccgx_hpi_device_reg_write() instead of fu_ccgx_hpi_device_i2c_write()